File: Source/Manager/LegacySPiDAccessToken/SPiDKeychainWrapper.m
//
//  SPiDKeychainWrapper.h
//  SPiDSDK
//
//  Copyright (c) 2012 Schibsted Payment. All rights reserved.
//

#import "SPiDKeychainWrapper.h"

@interface SPiDKeychainWrapper ()

/** Generates a service name to use for the keychain

 The service will have the form 'bundleIdentifier-SPiD'

 @return Service name
 */
+ (NSString *)serviceNameForSPiD;

/** Creates the basic search query used for all keychain operations

 @param identifier Unique identifier for the keychain item
 @return Query as a `NSMutableDictionary`
  */
+ (NSMutableDictionary *)setupSearchQueryForIdentifier:(NSString *)identifier;

@end

@implementation SPiDKeychainWrapper

#pragma mark Public methods

///---------------------------------------------------------------------------------------
/// @name Public methods
///---------------------------------------------------------------------------------------

+ (SPiDAccessToken *)accessTokenFromKeychainForIdentifier:(NSString *)identifier; {
    NSMutableDictionary *query = [self setupSearchQueryForIdentifier:identifier];

    // search attributes
    [query setObject:(__bridge id) kCFBooleanTrue forKey:(__bridge id) kSecMatchLimitOne];
    [query setObject:(__bridge id) kCFBooleanTrue forKey:(__bridge id) kSecReturnData];

    CFTypeRef cfData = nil;
    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef) query, &cfData);
    if (status == noErr) {
        NSData *result = (__bridge_transfer NSData *) cfData;
        SPiDAccessToken *accessToken = [NSKeyedUnarchiver unarchiveObjectWithData:result];
        return accessToken;
    } else {
        //NSAssert(status == errSecItemNotFound, @"Error reading from keychain");
        return nil;
    }
}

+ (BOOL)storeInKeychainAccessTokenWithValue:(SPiDAccessToken *)accessToken forIdentifier:(NSString *)identifier {
    NSData *data = [NSKeyedArchiver archivedDataWithRootObject:accessToken];
    NSMutableDictionary *query = [self setupSearchQueryForIdentifier:identifier];

    // add data
    [query setObject:data forKey:(__bridge id) kSecValueData];

    OSStatus status = SecItemAdd((__bridge CFDictionaryRef) query, NULL);
    if (status == errSecSuccess) {
        return YES;
    } else if (status == errSecDuplicateItem) {
        return [self updateAccessTokenInKeychainWithValue:accessToken forIdentifier:identifier];
    } else {
        // TODO: should we throw error instead?
        return NO;
    }
}

+ (BOOL)updateAccessTokenInKeychainWithValue:(SPiDAccessToken *)accessToken forIdentifier:(NSString *)identifier {
    NSData *data = [NSKeyedArchiver archivedDataWithRootObject:accessToken];
    NSMutableDictionary *searchQuery = [self setupSearchQueryForIdentifier:identifier];
    NSMutableDictionary *updateQuery = [[NSMutableDictionary alloc] init];

    // add data
    [updateQuery setObject:data forKey:(__bridge id) kSecValueData];

    OSStatus status = SecItemUpdate((__bridge CFDictionaryRef) searchQuery, (__bridge CFDictionaryRef) updateQuery);
    if (status == errSecSuccess) {
        return YES;
    } else {
        return NO;
    }
}

+ (void)removeAccessTokenFromKeychainForIdentifier:(NSString *)identifier {
    NSMutableDictionary *query = [self setupSearchQueryForIdentifier:identifier];

    OSStatus status = SecItemDelete((__bridge CFDictionaryRef) query);
    if (status != noErr) {
        SPiDDebugLog(@"Error deleting item to keychain - %d", (int)status);
    }
}

#pragma mark Private methods

///---------------------------------------------------------------------------------------
/// @name Private methods
///---------------------------------------------------------------------------------------

+ (NSString *)serviceNameForSPiD {
    NSString *appName = [[NSBundle mainBundle] bundleIdentifier];
    return [NSString stringWithFormat:@"%@-SPiD", appName];
}

+ (NSMutableDictionary *)setupSearchQueryForIdentifier:(NSString *)identifier {
    NSMutableDictionary *query = [[NSMutableDictionary alloc] init];

    [query setObject:(__bridge id) kSecClassGenericPassword forKey:(__bridge id) kSecClass];

    // set unique identification
    [query setObject:identifier forKey:(__bridge id) kSecAttrGeneric];
    [query setObject:identifier forKey:(__bridge id) kSecAttrAccount];
    [query setObject:[self serviceNameForSPiD] forKey:(__bridge id) kSecAttrService];

    return query;
}

@end

File: Source/Manager/LegacySPiDAccessToken/SPiDKeychainWrapper.swift
//
// Copyright 2011 - 2020 Schibsted Products & Technology AS.
// Licensed under the terms of the MIT license. See LICENSE in the project root.
//

import Foundation

final class SPiDKeychainWrapper {
    public class func accessTokenFromKeychain(forIdentifier identifier: String) -> SPiDAccessToken! {
        var query = setupSearchQuery(identifier: identifier)
        query[(kSecMatchLimitOne as String)] = true
        query[(kSecReturnData as String)] = true

        var cfData: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &cfData)
        if status == noErr, let data = cfData as? Data {
            if let accessToken = NSKeyedUnarchiver.unarchiveObject(with: data) as? SPiDAccessToken {
                return accessToken
            }
        }

        return nil
    }

    @discardableResult
    public class func storeInKeychainAccessToken(withValue accessToken: SPiDAccessToken, forIdentifier identifier: String) -> Bool {
        let data = NSKeyedArchiver.archivedData(withRootObject: accessToken)
        var query = setupSearchQuery(identifier: identifier)
        query[(kSecValueData as String)] = data

        let status = SecItemAdd(query as CFDictionary, nil)
        if status == errSecSuccess {
            return true
        } else if status == errSecDuplicateItem {
            return updateAccessTokenInKeychain(value: accessToken, forIdentifier: identifier)
        }

        return false
    }

    @discardableResult
    public class func updateAccessTokenInKeychain(value accessToken: SPiDAccessToken, forIdentifier identifier: String) -> Bool {
        let data = NSKeyedArchiver.archivedData(withRootObject: accessToken)
        let searchQuery = setupSearchQuery(identifier: identifier)
        let updateQuery: [String: Any] = [(kSecValueData as String): data]

        let status = SecItemUpdate(searchQuery as CFDictionary, updateQuery as CFDictionary)
        return status == errSecSuccess
    }

    public class func removeAccessTokenFromKeychain(forIdentifier identifier: String) {
        let query = setupSearchQuery(identifier: identifier)
        let status = SecItemDelete(query as CFDictionary)
        if status != noErr {
            log("Error deleting item to keychain - \(status)")
        }
    }

    private class var serviceNameForSPiD: String {
        let appName = Bundle.main.bundleIdentifier!
        return "\(appName)-SPiD"
    }

    private class func setupSearchQuery(identifier: String) -> [String: Any] {
        return [(kSecClass as String): kSecClassGenericPassword,
                (kSecAttrGeneric as String): identifier,
                (kSecAttrAccount as String): identifier,
                (kSecAttrService as String): serviceNameForSPiD]
    }
}
